README.md
# Instant18n Gem for Rails

Use OpenAI's GPT large-language model to power internationalization of the text in your Rails application. Extracted from real-world usage in [MagmaChat](https://github.com/magma-labs/magma-chat).

## Installation

Install the gem and add to the application's Gemfile by executing:

    $ bundle add instant18n

If bundler is not being used to manage dependencies, install the gem by executing:

    $ gem install instant18n

Make sure to set `OPENAI_ACCESS_TOKEN` in your environment so that the library is able to access GPT.

## Usage

Invoke with `I18n.it` or simply `it` in your view templates. (Method is short for _instant translation_). Use in place of the standard `t` method for translating text.

The `it` method provides translation using the GPT-3 language model and caching the results to improve performance.

```
>> I18n.it('Hello world!', 'español')
=> Hola mundo!
```

This will attempt to translate the text "Hello world!" to Spanish using the GPT-3 language model. If the translation is successful, the translated text will be returned. If the translation fails, the original text (or GPT error) will be returned.

### Options

`I18n.it(text, lang, opts)`

The `it` method accepts the following parameters:

- key (required): The key associated with the text to be translated.
- lang (required): The language to translate the text to. Defaults to the default language set in the I18n module.
- class: if you pass in css classes with the `class` option, the method will return the translation wrapped in a `div` tag, instead of plain text.

Additional options that affect caching:
- force: force a cache miss
- expires_in: (seconds) how long to cache the translation

Additional options that are passed to the GPT-3 API:

- model: defaults to gpt-3.5-turbo
- temperature: defaults to 0.25
- max_tokens: defaults to 64
- top_p: defaults to 0.1
- frequency_penalty: defaults to 0
- presence_penalty: defaults to 0

Full description of these options is available [here](https://platform.openai.com/docs/api-reference/chat/create).

### View Helper

`it(text, opts)`

This gem mixes in an `it` helper method into `ActionView::Base`. For convenience, the helper method assumes the presence of a `current_user` object with a `preferred_language` attribute. If `current_user` is nil, it will use the value of `I18n.default_language` instead.

### Default Language

The default language is set to `English`. For performance and practical reasons, if you pass in the default language, GPT is not invoked. Change the default language in an initializer or at runtime by changing the value of the `default_language` property on the `I18n` module.

```
I18n.default_language = "Spanish"
```

### Anything Goes

Because GPT is smart and can translate into almost anything that resembles a language, all of the following options are known to work:
* Español
* Baby Talk
* Baseldeutsch
* Braille
* Ebonics
* Emoji
* Esperanto
* Gregg Shorthand
* हिन्दी
* 日本語
* Klingon
* 1337 Speak (Leetspeak)
* 한국어
* 中文
* Newspeak
* Morse Code
* Rhyming Cockney Slang
* Sindarin
* Singlish
* Spanglish
* العربية
* Trumpisms
* Türkçe
* Uwu

The limit is your imagination!
## Development

After checking out the repo, run `bin/setup` to install dependencies. Then, run `rake spec` to run the tests. You can also run `bin/console` for an interactive prompt that will allow you to experiment.

To install this gem onto your local machine, run `bundle exec rake install`. To release a new version, update the version number in `version.rb`, and then run `bundle exec rake release`, which will create a git tag for the version, push git commits and the created tag, and push the `.gem` file to [rubygems.org](https://rubygems.org).

## Testing

The I18n Extensions gem can be tested using the RSpec testing framework. The tests are located in the spec directory and can be run using the following command:

```
bundle exec rspec
```

## Contributing

Bug reports and pull requests are welcome on GitHub at https://github.com/obie/instant18n. This project is intended to be a safe, welcoming space for collaboration, and contributors are expected to adhere to the [code of conduct](https://github.com/obie/instant18n/blob/main/CODE_OF_CONDUCT.md).

## License

The gem is available as open source under the terms of the [MIT License](https://opensource.org/licenses/MIT).

## Code of Conduct

Everyone interacting in the Instant18n project's codebases, issue trackers, chat rooms and mailing lists is expected to follow the [code of conduct](https://github.com/obie/instant18n/blob/main/CODE_OF_CONDUCT.md).

## Acknowledgments

The I18n Extensions gem uses the GPT-3 language model API provided by OpenAI.


CHANGELOG.md
## [Unreleased]

## [0.1.0] - 2023-04-28

- Initial release


spec/spec_helper.rb
# frozen_string_literal: true

require "instant18n"

RSpec.configure do |config|
  # Enable flags like --only-failures and --next-failure
  config.example_status_persistence_file_path = ".rspec_status"

  # Disable RSpec exposing methods globally on `Module` and `main`
  config.disable_monkey_patching!

  config.expect_with :rspec do |c|
    c.syntax = :expect
  end
end


spec/instant18n_spec.rb
# frozen_string_literal: true
require "dotenv/load"
require "rails"

class TestApp < Rails::Application
  config.cache_store = :memory_store
  config.eager_load = false
end

Rails.application.initialize!

RSpec.describe Instant18n do
  it "has a version number" do
    expect(Instant18n::VERSION).not_to be nil
  end

  it "does a translation", :aggregate_failures do
    expect(I18n.it("Save", "Español")).to eq("Guardar")
    expect(I18n.it("Save", "Pig Latin")).to eq("Avesay")
    expect(I18n.it("Save", "Emoji")).to eq("💾")
  end

  it "generates a translation in a div", :aggregate_failures do
    expect(I18n.it("Save", "Español", class: "foo bar")).to eq("<div class=\"instant18n espanol foo bar\">Guardar</div>")
  end

  it "does not call OpenAI when the language is English (default)", :aggregate_failures do
    # no instance of OpenAI::Client should be created or called
    expect(OpenAI::Client).not_to receive(:new)
    expect(I18n.it("Save", "English")).to eq("Save")
    expect(I18n.it("Save", "English", class: "foo bar")).to eq("<div class=\"instant18n english foo bar\">Save</div>")
  end
end


lib/i18n_extensions.rb
require "action_view/helpers/tag_helper"
require "openai"
require "rails"

module I18n
  extend ActionView::Helpers::TagHelper

  DIRECTIVE = "You are not a helper anymore, you are now a reliable translation web service."

  class << self
    attr_accessor :default_language
  end

  self.default_language = "English"

  ##
  # This method, `it`, instantly translates a given text based on the provided language code.
  # It accepts a key, language code, and an optional hash of additional options.
  #
  # @param key [String] the text to be translated.
  # @param lang [String] the target language or language code for translation (e.g., 'English', 'Español', 'fr').
  # @param options [Hash] optional hash of additional configuration options.
  #   - :class [Array<String>] an array of CSS class names to be applied to the translated
  # text when wrapped in a div.
  #   - :expires_in [Integer] the number of seconds to cache the translation.
  #   - :force [Boolean] whether to force a cache miss and re-translate the text.
  #   - :model [String] the name of the GPT-3 language model to use for translation.
  #   - :temperature [Float] the temperature to use for translation.
  #   - :max_tokens [Integer] the maximum number of tokens to use for translation.
  #   - :top_p [Float] the top p value to use for translation.
  #   - :frequency_penalty [Float] the frequency penalty to use for translation.
  #   - :presence_penalty [Float] the presence penalty to use for translation.
  #
  # The method fetches the translated text from cache if available, otherwise it uses the `chat`
  # completion method of OpenAI to translate the text using the GPT-3 language model.
  # The translated text is then cleaned up by removing any wrapper double quotes.
  # If the :classes option is provided, the translated text is wrapped in a div with the
  # specified css classes (in addition to `instant18n` and the language supplied, i.e. `espanol`).
  # Note that the language will be parameterized, so that `Español` becomes `espanol`.
  #
  # @return [String] the translated text.
  ##
  def self.it(key, lang, **opts)
    opts && opts.symbolize_keys!
    Rails.cache.fetch(cache_key(key, lang, opts), expires_in: (opts[:expires_in] || 1.year) , force: (opts[:force] || false)) do
      if lang.casecmp(default_language).zero?
        key
      else
        chat(prompt % { key: key, lang: lang }, directive: DIRECTIVE, **opts)
      end
    end.then do |text|
      text = text.to_s.gsub(/^"+|"+$/, '') # remove wrapper double quotes
      if opts[:class].present?
        tag.div(text, class: "instant18n #{lang.parameterize} #{opts[:class]}")
      else
        text
      end
    end
  end

  private

  def self.cache_key(key, lang, options)
    "instant18n_#{key}_#{lang}_#{options.values.join("_")}"
  end

  def self.chat(prompt, directive:, **options)
    messages = [ gpt_message(:system, directive) ]
    messages += [ gpt_message(:user, prompt) ]

    params = {
      messages: messages,
      model: options[:model] || "gpt-3.5-turbo",
      temperature: options[:temperature] || 0.25,
      max_tokens: options[:max_tokens] || 64,
      top_p: options[:top_p] || 0.1,
      frequency_penalty: options[:frequency_penalty] || 0,
      presence_penalty: options[:presence_penalty] || 0,
    }

    openai_client.chat(parameters: params).then do |response|
      response.dig("error","message") || response.dig("choices", 0, "message", "content")
    end
  end

  def self.prompt
    <<-END
    How do you write the text "%{key}" in clean, well-written %{lang} suitable for use as copy in a web application?
    Do not add any commentary or punctuation! Only respond with the answer, unquoted. If your response is just a single word
    or short phrase then do not add extra punctuation. The length of the response should match the length
    of the prompt. If you have very low confidence in the translation, then just respond with the original text.
    END
  end

  def self.openai_client
    @client ||= OpenAI::Client.new(access_token: ENV.fetch("OPENAI_ACCESS_TOKEN"))
  end

  def self.gpt_message(role, content)
    { role: role.to_s, content: content }
  end
end
